cabal-install/cabal-install.cabal

```diff
@@ -189,6 +189,11 @@ Flag network-uri
   description:  Get Network.URI from the network-uri package
   default:      True
 
+Flag debug-assertions
```

Comment:
I like the names `debug-expensive-assertions` and `expensiveAssert`.  I think that it would be simpler to use `ghc-options: -fno-ignore-asserts` to enable the usual assertions in cabal.project, though, because we could set the same flag for all packages: https://github.com/haskell/cabal/blob/9348682e939758a35d11534e6a723b799cbada53/cabal.project#L11